jubatus/util/math/random/sfmt.h

```diff
@@ -0,0 +1,250 @@
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+#ifndef JUBATUS_UTIL_MATH_RANDOM_SFMT_H_
+#define JUBATUS_UTIL_MATH_RANDOM_SFMT_H_
+#include <cstring>
+#include <stdint.h>
+
+#ifdef __SSE2__
+#include <emmintrin.h>
+#else
+typedef uint32_t __m128i __attribute__ ((__vector_size__ (16)));
+#endif
+
+namespace jubatus {
+namespace util{
+namespace math{
+namespace random{
+
+template<uint32_t MEXP, int POS1, int SL1, int SL2, int SR1, int SR2,
+         uint32_t MSK1, uint32_t MSK2, uint32_t MSK3, uint32_t MSK4,
+         uint32_t PARITY1, uint32_t PARITY2, uint32_t PARITY3, uint32_t PARITY4>
+class sfmt {
+  static const uint32_t N = MEXP / 128 + 1;
+  static const uint32_t N32 = N * 4;
+  union {
+    uint32_t u[4];
+    uint64_t u64[2];
+    __m128i si;
+  } s[N];
+  uint64_t idx;
+
+public:
+  sfmt(uint32_t seed) : idx(N32) {
+    static const uint32_t parity[4] = {PARITY1, PARITY2, PARITY3, PARITY4};
+    uint32_t *psfmt32 = &(s[0].u[0]);
+    psfmt32[0] = seed;
+    for (uint32_t i = 1; i < N32; i++) {
+      psfmt32[i] = 1812433253UL * (psfmt32[i - 1] ^ (psfmt32[i - 1] >> 30)) + i;
+    }
+
+    //period_certification
+    int inner = 0;
+    for (int i = 0; i < 4; i++)
+      inner ^= psfmt32[i] & parity[i];
+    for (int i = 16; i > 0; i >>= 1)
+      inner ^= inner >> i;
+    inner &= 1;
+    if (inner == 1)
+      return;
+    for (int i = 0; i < 4; i++) {
+      int work = 1;
+      for (int j = 0; j < 32; j++) {
+        if ((work & parity[i]) != 0) {
+          psfmt32[i] ^= work;
+          return;
+        }
+        work = work << 1;
+      }
+    }
+  }
+  ~sfmt() {}
+
+  uint32_t next() {
+    uint32_t *psfmt32 = &(s[0].u[0]);
+    if (idx >= N32) {
+      generate_all();
+      idx = 0;
+    }
+    return psfmt32[idx++];
+  }
+
+  void fill_int_unsafe(uint32_t *out, size_t n) {
+    uint32_t *psfmt32 = &(s[0].u[0]);
+    if (idx + n <= N32) {
+      std::memcpy(out, psfmt32 + idx, n * sizeof(uint32_t));
+      idx += n;
+    } else {
+      while (n > 0) {
+        if (idx >= N32) {
+          generate_all();
+          idx = 0;
+        }
+        size_t m = (n + idx <= N32 ? n : N32 - idx);
+        std::memcpy(out, psfmt32 + idx, m * sizeof(uint32_t));
+        idx += m;
+        out += m;
+        n -= m;
+      }
+    }
+  }
+private:
+  inline static void mm_recursion(__m128i * r, __m128i a, __m128i b,
+                                  __m128i c, __m128i d) {
+#ifdef __SSE2__
+    static const uint32_t sse2_param_mask[4] __attribute__((aligned(16)))
+      = {MSK1, MSK2, MSK3, MSK4};
+    __m128i v, x, y, z;
+    y = _mm_srli_epi32(b, SR1);
+    z = _mm_srli_si128(c, SR2);
+    v = _mm_slli_epi32(d, SL1);
+    z = _mm_xor_si128(z, a);
+    z = _mm_xor_si128(z, v);
+    x = _mm_slli_si128(a, SL2);
+    y = _mm_and_si128(y, *(const __m128i*)sse2_param_mask);
+    z = _mm_xor_si128(z, x);
+    z = _mm_xor_si128(z, y);
+    *r = z;
+#else
+    __m128i x, y;
```

Comment:
__m128i defined at L67 but raised compile error in non-SSE2 environment.
Thanks ! I will fix it.